identity/handler.go

```diff
@@ -675,3 +682,112 @@ func (h *Handler) patch(w http.ResponseWriter, r *http.Request, ps httprouter.Pa
 
 	h.r.Writer().Write(w, r, WithCredentialsMetadataAndAdminMetadataInJSON(*identity))
 }
+
+// Delete Credential Parameters
+//
+// swagger:parameters deleteIdentityCredentials
+// nolint:deadcode,unused
+type deleteIdentityCredentials struct {
+	// ID is the identity's ID.
+	//
+	// required: true
+	// in: path
+	ID string `json:"id"`
+
+	// Type is the credential's Type.
+	// One of totp, webauthn, lookup
+	//
+	// enum: totp,webauthn,lookup
+	// required: true
+	// in: path
+	Type string `json:"type"`
+}
+
+// swagger:route DELETE /admin/identities/{id}/credentials/{type} identity deleteIdentityCredentials
+//
+// # Delete a credential for a specific identity
+//
+// Delete an [identity](https://www.ory.sh/docs/kratos/concepts/identity-user-model) credential by its type
+// You can only delete second factor (aal2) credentials.
+//
+//	Consumes:
+//	- application/json
+//
+//	Produces:
+//	- application/json
+//
+//	Schemes: http, https
+//
+//	Security:
+//	  oryAccessToken:
+//
+//	Responses:
+//	  200: identity
+//	  404: errorGeneric
+//	  default: errorGeneric
+func (h *Handler) deleteIdentityCredentials(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	identity, err := h.r.PrivilegedIdentityPool().GetIdentityConfidential(r.Context(), x.ParseUUID(ps.ByName("id")))
+	if err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	cred, ok := identity.GetCredentials(CredentialsType(ps.ByName("type")))
+	if !ok {
+		h.r.Writer().WriteError(w, r, errors.WithStack(herodot.ErrNotFound.WithReasonf("You tried to remove a %s but this user have no %s set up.", ps.ByName("type"), ps.ByName("type"))))
+		return
+	}
+
+	switch cred.Type {
+	case CredentialsTypeLookup:
+		fallthrough
+	case CredentialsTypeTOTP:
+		identity.DeleteCredentialsType(cred.Type)
+	case CredentialsTypeWebAuthn:
+		var cc CredentialsWebAuthnConfig
+		if err := json.Unmarshal(cred.Config, &cc); err != nil {
+			h.r.Writer().WriteError(w, r, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to decode identity credentials.").WithDebug(err.Error())))
+			return
+		}
+
+		updated := make([]CredentialWebAuthn, 0)
+		for k, cred := range cc.Credentials {
+			if cred.IsPasswordless {
+				updated = append(updated, cc.Credentials[k])
+			}
+		}
+
+		if len(updated) == 0 {
+			identity.DeleteCredentialsType(CredentialsTypeWebAuthn)
+			break
+		}
+
+		cc.Credentials = updated
+		cred.Config, err = json.Marshal(cc)
+		if err != nil {
+			h.r.Writer().WriteError(w, r, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to encode identity credentials.").WithDebug(err.Error())))
+			return
+		}
+
+		identity.SetCredentials(CredentialsTypeWebAuthn, *cred)
```

Comment:
This is missing one test in the test suite, having webauthn credentials with both passwordless and 2fa credentials, and then testing that we do indeed get the correct result:

**before**
* webauthn key 1 passwordless
* webauthn key 2 passwordless
* webauthn key 3 2fa
* webauthn key 4 2fa

**expected result**
* webauthn key 3 2fa
* webauthn key 4 2fa

My recommendation would be to move this code into  a separate function, write unit tests there. This will make the testing in handler_test a bit easier

Reply:
I did move the webauthn removal to his own function but I didn't tested that function per-se as it was complexe to build and my test function above had all the bits need to test the webauthn removal;

(If I want to keep the `deletCredentialWebAuthFromIdentity` private I would need to trick the testing to code to make it public as the handler_test has it own pkg name)

But if needed I can write the code.

Reply:
I did move the webauthn removal to his own function but I didn't tested that function per-se as it was complexe to build and my test function above had all the bits need to test the webauthn removal;

(If I want to keep the `deletCredentialWebAuthFromIdentity` private I would need to trick the testing to code to make it public as the handler_test has it own pkg name)

But if needed I can write the code.

Reply:
you can create a new file `handler_unit_test`, use the proper package there, and write the test :) But the handler test now is sufficient, so it's OK imo :)